sentinelhub/sentinelhub_catalog.py

```diff
@@ -0,0 +1,247 @@
+    `Catalog API documentation <https://docs.sentinel-hub.com/api/latest/api/catalog>`_.
+    """
+    def __init__(self, base_url=None, config=None):
+        """
+        :param base_url: A base URL of Sentinel Hub service specifying which service deployment to use. By default the
+            one in config object will be used.
+        :type base_url: str or None
+        :param config: A configuration object with required parameters `sh_client_id`, `sh_client_secret`, and
+            `sh_base_url` which will be used of authentication.
+        :type config: SHConfig or None
+        """
+        self.config = config or SHConfig()
+
+        base_url = base_url or self.config.sh_base_url
+        base_url = base_url.rstrip('/')
+        self.catalog_url = f'{base_url}/api/v1/catalog'
+
+        self.client = SentinelHubDownloadClient(config=self.config)
+
+    def get_info(self):
+        """ Provides the main information that define Sentinel Hub Catalog API
+
+        :return: A service payload with information
+        :rtype: dict
+        """
+        return self.client.get_json(self.catalog_url)
+
+    def get_conformance(self):
+        """ Get information about specifications that this API conforms to
+
+        :return: A service payload with information
+        :rtype: dict
+        """
+        url = f'{self.catalog_url}/conformance'
+        return self.client.get_json(url)
+
+    def get_collections(self):
+        """ Provides a list of collections that are available to a user
+
+        :return: A list of collections with information
+        :rtype: list(dict)
+        """
+        url = f'{self.catalog_url}/collections'
+        return self.client.get_json(url, use_session=True)['collections']
+
+    def get_collection_info(self, collection):
+        """ Provides information about given collection
+
+        :param collection: A data collection object or a collection ID
+        :type collection: DataCollection or str
+        :return: Information about a collection
+        :rtype: dict
+        """
+        collection_id = self._parse_collection_id(collection)
+        url = f'{self.catalog_url}/collections/{collection_id}'
+        return self.client.get_json(url, use_session=True)
+
+    def get_feature(self, collection, feature_id):
+        """ Provides information about a single feature in a collection
+
+        :param collection: A data collection object or a collection ID
+        :type collection: DataCollection or str
+        :param feature_id: A feature ID
+        :type feature_id: str
+        :return: Information about a feature in a collection
+        :rtype: dict
+        """
+        collection_id = self._parse_collection_id(collection)
+        url = f'{self.catalog_url}/collections/{collection_id}/items/{feature_id}'
+        return self.client.get_json(url, use_session=True)
+
+    def search(self, collection, *, time, bbox=None, geometry=None, ids=None, query=None, fields=None, distinct=None,
```

Comment:
In this case, a search query object would have a single purpose - to be given to the search method. There is no other method that would also need an exactly such search query and I see no other use-case for it.

I don't see how changing
```Python
catalog.search(...)
```
into 
```Python
from sentinelhub import CatalogSearchQuery

query = CatalogSearchQuery(...)
catalog.search(query)
```
simplifies code from a user's point of view and also the implementation of `CatalogSearchQuery` cannot be any less simple as what is currently in `Catalog.search` method.

In the case of BYOC interface, it makes sense to have objects for collection and tile because they can be passed to and obtained from multiple methods as well as be used for further analysis. But here `Catalog.search` and `CatalogSearchQuery` would be always tightly coupled.

IMO, these are the criteria in deciding if something should be a class or just a method/function.

A side note:
It might only make sense to have a more general `SearchQuery` class that could be used for Catalog, WFS, and Opensearch. But since each of the services has different unique parameters I'm not sure how feasible would that be in practice.

Reply:
The simplification is that one can have an object that she can pass around, update, reuse, etc. Additionally, passing a single object to the method is much more clear than a (million) parameters... 

As I said, that would be my personal preference and I am not willing to loose sleep over it.